### ISSUE TITLE
Problem in training Neural network

### ISSUE THREAD

**Issue by pattysan** [zueve/neurolab#25]
I am pretty new in using python and neurolab and I have a problem with the training of my feed forward neural network. I have built the net as following:
net = nl.net.newff([[-1,1]]*64, [60,1])
net.init()
testerr = net.train(InputT, TargetT, epochs=100, show=1)
and my target output is a vector between 0 and 4. When I use the nl.train.train_bfgs I have in the console:
testerr = net.train(InputT, TargetT, epochs=10, show=1)
Epoch: 1; Error: 55670.4462766;
Epoch: 2; Error: 55649.5;
As you can see, I fixed the number of epochs to 100 but it stops at the second epoch and after the test of the net with Netresults=net.sim(InputCross) I have as test output array a vector of 1 (totally wrong). If I use the other training functions I have the same output testing vector full of 1 but in that case during the training, the epochs reach the number that I set but the error displayed doesn't change between the epochs. The same if the target output vector is between -1 and 1. Any suggestion? Thank you very much!

**Comment by pattysan** [comment]
I think that the problem about  the vector of 1 when the target vector is between 0 and 4  is solved by #8 but I have the same problem also when the output is between -1 and 1 ( I didn' t use the norm function) and  I still don' t understand why the  number of  epochs stops before.

**Comment by zueve** [comment]
nl.train.train_bfgs use scipy.optimize.fmin_bfgs and it has another stop fumction. 
epochs=100 must guarantee that count of epochs will be less then 100 in this train function.

Repeat train with output is between -1 and 1 and reduce the number of neurons in the input layer. Test with 20, 25, 30, 35 neurons

**Comment by pattysan** [comment]
Thank you for your reply.I'll try as soon as possible. But why if I use train.gdx as training function  I have a nan error or a fixed training error?

**Comment by zueve** [comment]
I think, train_gdx can not escape from a local minimum. It has same parameters for get better result

        lr: float (defaults 0.01)
            learning rate
        lr_inc: float (default 1.05)
            Ratio to increase learning rate
        lr_dec: float (default 0.7)
            Ratio to decrease learning rate
        max_perf_inc:float (default 1.04)
            Maximum performance increase
        mc: float (default 0.9)
            Momentum constant

**Comment by pattysan** [comment]
Hello, @zueve  I can' t  reduce the number of input neurons (they are now more  than 64). 
Did you suggest to reduce the number of them because neurolab has solme troubles with numerous input neurons or just for a try? Thank you very much.

**Comment by zueve** [comment]
Hello @pattysan!
in your code:
> net = nl.net.newff([[-1,1]]*64, [60,1])

64 - is number of inputs
60 is number of neurons in input/hidden layer, change it for try.

**Comment by pattysan** [comment]
I changed the input vector and tried also  to use train_gdx fixing the parameter mc and lr but the error is always fixed  to  a value and doesn' t change during the training. I need lots of inputs for my neural network (between 72 and 150) and  I can' t figure out how can I solve this training problem. My inputs and outputs are between -1 and 1.

**Comment by zueve** [comment]
It is difficult to determine what the problem is without data. Perhaps traingdx not suitable for your application

**Comment by pattysan** [comment]
I have just one doubt: am I setting well the network?

net=[]
net = nl.net.newff([[-1,1]]*72, [30,1])
net.errorf = nl.error.SAE() 
net.trainf = nl.train.train_gdx
net.init()
testerr=[]
testerr = net.train(InputT, TargetT,epochs=100, show=1,lr=0.5,adapt=True,mc=0.7)

In case it is ok, I guess that the problem is for my data or maybe too many inputs.
Thanks

**Comment by zueve** [comment]
I can not see serious problems. Why *adapt=True*? 
what is the result on this same network with *train_bfgs*?

**Comment by pattysan** [comment]
With  train_bfgs I have a memory issue (I am using a 32 bit computer).  
But still, is it correct how I wrote: net.trainf = nl.train.train_gdx  
that is I didn' t add the lr, mc and the other parameters here but in:   testerr = net.train(InputT, TargetT,epochs=100, show=1,lr=0.5,adapt=True,mc=0.7)
Thank you

**Comment by zueve** [comment]
> net.trainf = nl.train.train_gdx

is correct

**Comment by kobbe** [comment]
Hi, I had the same problem, it took about 5hr to find the problem, only found the stackoverflow problem page and not this one. 

Please make sure neurolab reports back the right status when fmin_bfgs stops in advance! It is very confusing that neurolab sometimes do not say that traning is complete.

**Comment by pattysan** [comment]
hI @kobbe ? Where did you find the solution of the problem?

**Comment by solarjoe** [comment]
[Docs](https://pythonhosted.org/neurolab/intro.html#support-neural-networks-types) mentions neurolab.train.train_gdx() as default for net.newff

[Source](https://pythonhosted.org/neurolab/_modules/neurolab/net.html) shows that train.train_bfgs is set as default.

**Comment by zueve** [comment]
Yes, i changed it few version ago...
